### PR TITLE
Refactor torch.*solve tests

### PR DESCRIPTION
**Pull request by vishwakftw**
Changelog:
- De-duplicate the code in tests for torch.solve, torch.cholesky_solve, torch.triangular_solve
- Skip tests explicitly if requirements aren't met for e.g., if NumPy / SciPy aren't available in the environment
- Add generic helpers for these tests in test/common_utils.py

Test Plan:
- All tests should pass to confirm that the change is not erroneous

Clears one point specified in the discussion in #24333.

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -5969,129 +5908,60 @@ def test_geqrf(self):
...
-            # basic correctness test
-            A, b = triangular_solve_test_helper((3, 5, 5), (3, 5, 10), cast, upper, unitriangular)
-            x = torch.triangular_solve(b, A, upper=upper, transpose=transpose, unitriangular=unitriangular)[0]
+            for i in range(b_dims[0]):
+                x_exp_list.append(torch.triangular_solve(b[i], A[i], upper=upper,
+                                                         unitriangular=unitriangular, transpose=transpose)[0])
+            x_exp = torch.stack(x_exp_list)  # Stacked output
+            x_act = torch.triangular_solve(b, A, upper=upper,
+                                           unitriangular=unitriangular, transpose=transpose)[0]  # Actual output
+            self.assertEqual(x_exp, x_act)  # Equality check
```
Comment (zou3519) [inline test/test_torch.py]: nit: switch the order of the args. Usually equality checks are `self.assertEqual(obtained_result, expected)`

File: test/test_torch.py [hunk 2]
```diff
@@ -6124,14 +5992,9 @@ def scipy_tri_solve_batched(A, B, upper, trans, diag):
...
 
         def run_test(A_dims, b_dims, cast, upper, transpose, unitriangular):
-            A = torch.randn(*A_dims)
-            A = A.triu() if upper else A.tril()
-            if unitriangular:
-                A.diagonal(dim1=-2, dim2=-1).fill_(1.)
-            b = torch.randn(*b_dims)
-            x_exp = torch.Tensor(scipy_tri_solve_batched(A.numpy(), b.numpy(),
+            b, A = triangular_solve_test_helper(A_dims, b_dims, cast, upper, unitriangular)
+            x_exp = torch.Tensor(scipy_tri_solve_batched(A.cpu().numpy(), b.cpu().numpy(),
```
Comment (zou3519) [inline test/test_torch.py]: Don't use the legacy `torch.Tensor` constructor, use `torch.tensor`

File: test/test_torch.py [hunk 3]
```diff
@@ -7447,75 +7311,44 @@ def test_cholesky_batched_many_batches(self):
...
-        # default arg Test
-        L_def = torch.cholesky(a)
-        x_def = torch.cholesky_solve(b, L_def)
-        self.assertLessEqual(b.dist(torch.mm(a, x_def)), 1e-12)
+        from common_utils import cholesky_solve_test_helper
+        for (k, n), upper in product(zip([2, 3, 5], [3, 5, 7]), [True, False]):
+            b, A, L = cholesky_solve_test_helper((n,), (n, k), cast, upper)
+            x = torch.cholesky_solve(b, L, upper=upper)
+            b_ = torch.matmul(A, x)
+            self.assertEqual(b_, b)
```
Comment (zou3519) [inline test/test_torch.py]: In the original test we are checking that the delta is less than 1e-12, should we keep that?

File: test/test_torch.py [hunk 4]
```diff
@@ -7447,75 +7311,44 @@ def test_cholesky_batched_many_batches(self):
...
             x_exp_list = []
-            for i in range(4):
-                x_exp = torch.cholesky_solve(b[i], L[i], upper=upper)
-                x_exp_list.append(x_exp)
-            x_exp = torch.stack(x_exp_list)
+            for i in range(b_dims[0]):
+                x_exp_list.append(torch.cholesky_solve(b[i], L[i], upper=upper))
+            x_exp = torch.stack(x_exp_list)  # Stacked output
+            x_act = torch.cholesky_solve(b, L, upper=upper)  # Actual output
+            self.assertEqual(x_exp, x_act)  # Equality check
```
Comment (zou3519) [inline test/test_torch.py]: nit: switch the order of the args in assertEqual

File: test/test_torch.py (hunk 3 above)
Comment (vishwakftw) [reply test/test_torch.py]: I don't think it would matter all that much.

File: test/test_torch.py (hunk 4 above)
Comment (vishwakftw) [reply test/test_torch.py]: Sure, thank you.

File: test/test_torch.py (hunk 1 above)
Comment (vishwakftw) [reply test/test_torch.py]: Sure, thank you.

File: test/test_torch.py (hunk 2 above)
Comment (vishwakftw) [reply test/test_torch.py]: I keep forgetting :-(

File: test/test_torch.py (hunk 3 above)
Comment (zou3519) [reply test/test_torch.py]: I don't have context on why the original error bound was 1e-12 so I'm a little hesitant to remove it. Does the test still pass with the 1e-12?

Comment (vishwakftw) [reply test/test_torch.py]: I’ll re-add it to the tests, push and await CI signal. Hope that is fine.

**Comment by vishwakftw** [comment]
Failures are unrelated, and previously failing ROCm test passed now

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**